gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/jwt/impl/JWTServiceImpl.java

```diff
@@ -92,31 +92,52 @@ public Single<String> encodeAuthorization(JWT jwt, Client client) {
     }
 
     @Override
-    public Single<JWT> decodeAndVerify(String jwt, Client client) {
+    public Single<JWT> decodeAndVerify(String jwt, Client client, TokenType tokenType) {
         return certificateManager.get(client.getCertificate())
                 .defaultIfEmpty(certificateManager.defaultCertificateProvider())
-                .flatMapSingle(certificateProvider -> decodeAndVerify(jwt, certificateProvider));
+                .flatMapSingle(certificateProvider -> decodeAndVerify(jwt, certificateProvider, tokenType));
     }
 
     @Override
-    public Single<JWT> decodeAndVerify(String jwt, CertificateProvider certificateProvider) {
-        return decode(certificateProvider, jwt)
+    public Single<JWT> decodeAndVerify(String jwt, CertificateProvider certificateProvider, TokenType tokenType) {
+        return decode(certificateProvider, jwt, tokenType)
                 .map(JWT::new);
     }
 
     @Override
-    public Single<JWT> decode(String jwt) {
+    public Single<JWT> decode(String jwt, TokenType tokenType) {
         return Single.create(emitter -> {
             try {
                 String json = new String(Base64.getDecoder().decode(jwt.split("\\.")[1]), "UTF-8");
                 emitter.onSuccess(objectMapper.readValue(json, JWT.class));
             } catch (Exception ex) {
-                logger.debug("Failed to decode JWT", ex);
-                emitter.onError(new InvalidTokenException("The access token is invalid", ex));
+                logger.debug("Failed to decode {} JWT", tokenType, ex);
+                emitter.onError(buildInvalidTokenException(tokenType, ex));
             }
         });
     }
 
+    private static InvalidTokenException buildInvalidTokenException(TokenType tokenType, Exception ex) {
+        var invalidToken = new InvalidTokenException("The access token is invalid", ex);
```

Comment:
Observation: it is always safer and cleaner no to reassign variable. No action required.